File: pe-core/src/main/java/org/saar/core/common/r3d/MeshWriter3D.java
package org.saar.core.common.r3d;

import org.joml.Matrix4fc;
import org.saar.core.mesh.writers.InstancedElementsMeshWriter;

public class MeshWriter3D implements InstancedElementsMeshWriter<Vertex3D, Instance3D> {

    private final MeshPrototype3D prototype;

    public MeshWriter3D(MeshPrototype3D prototype) {
        this.prototype = prototype;
    }

    @Override
    public void writeVertex(Vertex3D vertex) {
        this.prototype.getPositionBuffer().getWriter().write3f(vertex.getPosition3f());
        this.prototype.getNormalBuffer().getWriter().write3f(vertex.getNormal3f());
        this.prototype.getColourBuffer().getWriter().write3f(vertex.getColour3f());
    }

    @Override
    public void writeInstance(Instance3D instance) {
        final Matrix4fc matrix = instance.getTransform().getTransformationMatrix();
        this.prototype.getTransformBuffer().getWriter().write4x4f(matrix);
    }

    @Override
    public void writeIndex(int index) {
        this.prototype.getIndexBuffer().getWriter().writeInt(index);
    }

}

File: pe-core/src/main/java/org/saar/core/common/r3d/MeshWriter3D.kt
package org.saar.core.common.r3d

import org.saar.core.mesh.builder.IndexedMeshWriter
import org.saar.core.mesh.builder.InstancedMeshWriter
import org.saar.core.mesh.builder.VertexMeshWriter
import org.saar.lwjgl.util.DataWriter

class MeshWriter3D(
    private val positionWriter: DataWriter,
    private val normalWriter: DataWriter,
    private val colourWriter: DataWriter,
    private val transformWriter: DataWriter,
    private val indexWriter: DataWriter,
) : VertexMeshWriter<Vertex3D>, InstancedMeshWriter<Instance3D>, IndexedMeshWriter {

    override fun writeVertex(vertex: Vertex3D) {
        this.positionWriter.write3f(vertex.position3f)
        this.normalWriter.write3f(vertex.normal3f)
        this.colourWriter.write3f(vertex.colour3f)
    }

    override fun writeInstance(instance: Instance3D) {
        this.transformWriter.write4x4f(instance.transform.transformationMatrix)
    }

    override fun writeIndex(index: Int) = this.indexWriter.writeInt(index)
}